packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,203 @@
+          })
+        ).then(resolve, reject)
+      } catch (err) {
+        reject(err)
+      }
+    })
+  })
+}
+
+/**
+ *
+ * @param {AugmentedJob} job
+ * @param {Job["plugin"]} plugin
+ * @param {pDefer.DeferredPromise} deferred
+ */
+const runJob = (job, plugin, deferred) => {
+  const worker = require(`${plugin.name}/worker.js`)
+  if (!worker[job.name]) {
+    deferred.reject(new Error(`No worker function found for ${job.name}`))
+    return
+  }
+
+  runLocalWorker(worker[job.name], job).then(deferred.resolve, deferred.reject)
+}
+
+const handleJobEnded = () => {
+  if (--activeJobs === 0) {
+    activityForJobs.end()
+    activityForJobs = null
+  }
+}
+
+/**
+ * Creates a job
+ *
+ * @param {Job} args
+ * @return {Promise<unknown>}
+ */
+exports.enqueueJob = async ({ name, inputPaths, outputDir, args, plugin }) => {
+  const rootDir = store.getState().program.directory
+
+  // TODO see if we can make this async, filehashing might be expensive to wait for
+  // currently this needs to be sync as we could miss jobs to have been scheduled and
+  // are still processing their hashes
+  const inputPathsWithContentDigest = inputPaths.map(path => {
+    return {
+      path: convertPathsToRelative(path, rootDir),
+      contentDigest: createFileHash(path),
+    }
+  })
+
+  const job = {
+    id: uuid(),
+    name,
+    inputPaths: inputPathsWithContentDigest,
+    outputDir: convertPathsToRelative(outputDir, rootDir),
+    args,
+    plugin,
+  }
+
+  job.contentDigest = createContentDigest({
+    name: job.name,
+    inputPaths: job.inputPaths.map(inputPath => inputPath.contentDigest),
+    outputDir: job.outputDir,
+    args: job.args,
+    plugin: job.plugin,
+  })
+
+  // When we already have a job that's executing, return the same promise.
+  if (jobsInProcess.has(job.contentDigest)) {
+    return jobsInProcess.get(job.contentDigest).deferred.promise
+  }
+
+  // Bump active jobs
+  activeJobs++
+  if (!activityForJobs) {
+    activityForJobs = reporter.phantomActivity(`Running jobs`)
+    activityForJobs.start()
+  }
+
+  // TODO: check cache folder for already stored jobs
+  const deferred = pDefer()
+  // node 8 doenst have finally so we call then & catch
+  deferred.promise = deferred.promise
+    .catch(err => {
+      handleJobEnded()
+
+      throw err
+    })
+    .then(res => {
+      handleJobEnded()
+
+      return res
+    })
+
+  jobsInProcess.set(job.contentDigest, {
+    id: job.id,
+    deferred,
+  })
+
+  runJob(job, plugin, deferred)
+
+  return deferred.promise
+}
+
+/**
+ * Wait for all processing jobs to have finished
+ *
+ * @return {Promise<void>}
+ */
+exports.waitUntilAllJobsComplete = () => {
+  const jobsPromises = []
+  jobsInProcess.forEach(({ deferred }) => jobsPromises.push(deferred.promise))
+
+  return Promise.all(jobsPromises).then(() => {})
+}
+
+exports.resolveWorker = plugin => {
+  try {
+    return require.resolve(`${plugin.name}/worker.js`)
+  } catch (err) {
+    throw new Error(
```

Comment:
Would not recommend eating the actual error here without logging it anywhere. This could become a problem while debugging.

Reply:
Normally I would agree with this but the original error would say exactly the same, it will say that the worker does not exist. I don't see the need to let it throw the original error here.